File: pharmarobot/pharmarobot_ros2-master/src/kalman_odom/src/test_sim_input.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/vector3.hpp>
#include <geometry_msgs/msg/pose2_d.hpp>
#include <cmath>

class SimTestNode : public rclcpp::Node {
public:
    SimTestNode() : Node("sim_test_node") {
        wheel_pub_ = create_publisher<geometry_msgs::msg::Vector3>("/wheel_displacement", 10);
        lidar_pub_ = create_publisher<geometry_msgs::msg::Pose2D>("/lidar/pose2d", 10);
        timer_ = create_wall_timer(std::chrono::milliseconds(100), std::bind(&SimTestNode::publish_mock_data, this));
    }

private:
    rclcpp::Publisher<geometry_msgs::msg::Vector3>::SharedPtr wheel_pub_;
    rclcpp::Publisher<geometry_msgs::msg::Pose2D>::SharedPtr lidar_pub_;
    rclcpp::TimerBase::SharedPtr timer_;
    double t_ = 0.0;

    void publish_mock_data() {
        auto wheel_msg = geometry_msgs::msg::Vector3();
        wheel_msg.x = 0.01 + 0.002 * sin(t_);
        wheel_msg.y = 0.01 - 0.002 * sin(t_);
        wheel_pub_->publish(wheel_msg);

        if (static_cast<int>(t_ * 10) % 10 == 0) {  // simulate lidar at 1Hz
            auto lidar_msg = geometry_msgs::msg::Pose2D();
            lidar_msg.x = 0.1 * t_;
            lidar_msg.y = 0.05 * t_;
            lidar_msg.theta = 0.01 * t_;
            lidar_pub_->publish(lidar_msg);
        }

        t_ += 0.1;
    }
};

File: pharmarobot/pharmarobot_ros2-master/src/kalman_odom/scripts/test_sim_input.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Vector3, Pose2D
import math

class SimTestNode(Node):
    def __init__(self):
        super().__init__('sim_test_node')
        self.wheel_pub = self.create_publisher(Vector3, '/wheel_displacement', 10)
        self.lidar_pub = self.create_publisher(Pose2D, '/lidar/pose2d', 10)
        self.timer = self.create_timer(0.1, self.publish_mock_data)
        self.t = 0.0

    # straight line motion with slight oscillation
    # and occasional LIDAR correction
    # This simulates a robot moving forward with slight variations in wheel displacement
    # and provides LIDAR corrections at regular intervals.
    # The wheel displacements are simulated as small oscillations around a base value,
    # while the LIDAR corrections are provided at every 10th timer tick.
    
    def publish_straight_line_data(self):
        
        # generate random linear and angular velocities
        linear_velocity = 0.1 + 0.02 * math.sin(self.t)
        angular_velocity = 0.01 * math.cos(self.t)
        
        # Simulate wheel encoder displacement (in radians)
        dl = linear_velocity * 0.1 + 0.002 * math.sin(self.t)
        dr = linear_velocity * 0.1 - 0.002 * math.sin(self.t)   
        
    
    
    def publish_mock_data(self):
        # Simulate wheel encoder displacement (in radians)
        dl = 0.01 + 0.002 * math.sin(self.t)
        dr = 0.01 - 0.002 * math.sin(self.t)

        wheel_msg = Vector3()
        wheel_msg.x = dl
        wheel_msg.y = dr
        self.wheel_pub.publish(wheel_msg)

        # Simulate occasional LIDAR correction
        if int(self.t * 10) % 10 == 0:
            lidar_msg = Pose2D()
            lidar_msg.x = 0.1 * self.t
            lidar_msg.y = 0.05 * self.t
            lidar_msg.theta = 0.01 * self.t
            self.lidar_pub.publish(lidar_msg)

        self.t += 0.1


def main(args=None):
    rclpy.init(args=args)
    node = SimTestNode()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
